README.md
# autocrop
[![Travis Build Status](https://img.shields.io/travis/leblancfg/autocrop/master.svg)](https://travis-ci.org/leblancfg/autocrop) [![AppVeyor Build Status](https://img.shields.io/appveyor/ci/leblancfg/autocrop/master.svg?label=%22Windows%22)](https://ci.appveyor.com/project/leblancfg/autocrop/branch/master) [![Codecov master](https://img.shields.io/codecov/c/github/leblancfg/autocrop/master.svg)](https://codecov.io/gh/leblancfg/autocrop) [![PyPI version](https://badge.fury.io/py/autocrop.svg)](https://badge.fury.io/py/autocrop) [![Downloads](https://pepy.tech/badge/autocrop)](https://pepy.tech/project/autocrop) [![Known Vulnerabilities](https://snyk.io/test/github/leblancfg/autocrop/badge.svg?targetFile=requirements.txt)](https://snyk.io/test/github/leblancfg/autocrop?targetFile=requirements.txt) [![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/leblancfg/autocrop.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/leblancfg/autocrop/context:python)

<p align="center"><img title="obama_crop" src="https://cloud.githubusercontent.com/assets/15659410/10975709/3e38de48-83b6-11e5-8885-d95da758ca17.png"></p>

Perfect for profile picture processing for your website or batch work for ID cards, autocrop will output images centered around the biggest face detected.

## Use
From the command line:

	usage: [-h] [-o OUTPUT] [-i INPUT] [-w WIDTH] [-H HEIGHT] [-v]

	Automatically crops faces from batches of pictures

	optional arguments:
	  -h, --help
	  		Show this help message and exit
	  -o, --output, -p, --path
			Folder where cropped images will be placed.
			Default: current working directory
	  -r, --reject
			Folder where images without detected faces will be placed.
			Default: same as output directory
	  -i, --input
			Folder where images to crop are located.
			Default: current working directory
	  -w, --width
			Width of cropped files in px. Default=500
	  -H, --height
			Height of cropped files in px. Default=500
	  --facePercent
	  		Percentage of Face height to image height (zoom factor)
	  --padUp
	  		Padding up value compared to padDown. Default=50
	  --padDown
	  		Padding down value compared to padDown. Default=50
	  --padLeft
	  		Padding left value compared to padRight. Default=50
	  --padRight
	  		Padding right value compared to padLeft. Default=50
	  -v, --version
	  		Show program's version number and exit


### Examples

* Crop every image in the `pics` folder, resize them to 400 px squares, and output them in the `crop` directory:
	- `autocrop -i pics -o crop -w 400 -H 400`.
	- Images where a face can't be detected will be left in `crop`.
* Same as above, but output the images with undetected faces to the `reject` folder:
	- `autocrop -i pics -o crop -r nofaces -w 400 -H 400`.
* Same as the first example, but add more padding at the bottom:
	- `autocrop -i pics -o crop -w 400 -H 400 --facePercent 50 --padUp 20 --padDown 50`.
	
If no output folder is added, asks for confirmation and destructively crops images in-place.

### Supported file types

The following file types are supported:

- EPS files (`.eps`)
- GIF files (`.gif`) (only the first frame of an animated GIF is used)
- JPEG 2000 files (`.j2k`, `.j2p`, `.jp2`, `.jpx`)
- JPEG files (`.jpeg`, `.jpg`, `.jpe`)
- LabEye IM files (`.im`)
- macOS ICNS files (`.icns`)
- Microsoft Paint bitmap files (`.msp`)
- PCX files (`.pcx`)
- Portable Network Graphics (`.png`)
- Portable Pixmap files (`.pbm`, `.pgm`, `.ppm`)
- SGI files (`.sgi`)
- SPIDER files (`.spi`)
- TGA files (`.tga`)
- TIFF files (`.tif`, `.tiff`)
- WebP (`.webp`)
- Windows bitmap files (`.bmp`, `.dib`)
- Windows ICO files (`.ico`)
- X bitmap files (`.xbm`)

## Installation
Simple! In your command line, type:

~~~python
pip install autocrop
~~~

### Gotchas
Autocrop uses OpenCV to perform face detection, which is installed through binary [wheels](http://pythonwheels.com/). If you *already* have OpenCV 3+ installed, you may wish to uninstall the additional OpenCV installation: `pip uninstall opencv-python`.

### Installing directly
In some cases, you may wish the package directly, instead of through [PyPI](https://pypi.python.org/pypi):

~~~
cd ~
git clone https://github.com/leblancfg/autocrop
cd autocrop
pip install .
~~~

### conda
Development of a `conda-forge` package for the [Anaconda Python distribution](https://www.anaconda.com/download/) is also currently slated for development. Please leave feedback on [issue #7](https://github.com/leblancfg/autocrop/issues/7) if you are insterested in helping out.

## Requirements
Best practice for your projects is of course to [use virtual environments](http://docs.python-guide.org/en/latest/dev/virtualenvs/). At the very least, you will need to [have pip installed](https://pip.pypa.io/en/stable/installing/).

Autocrop is currently being tested on:
* Python:
    - 2.7
    - 3.4
    - 3.5
    - 3.6
* OS:
    - Linux
    - macOS
    - Windows

## More Info
Check out:
* http://docs.opencv.org/master/d7/d8b/tutorial_py_face_detection.html#gsc.tab=0
* http://docs.opencv.org/master/d5/daf/tutorial_py_histogram_equalization.html#gsc.tab=0

Adapted from:
* http://photo.stackexchange.com/questions/60411/how-can-i-batch-crop-based-on-face-location

## Contributing

Although autocrop is essentially a CLI wrapper around a single OpenCV function, it is actively developed. It has active users throughout the world.

If you would like to contribute, please consult the [contribution docs](CONTRIBUTING.md).


changelog.md
# Autocrop changelog

## 0.3.2
### Changes
* Autocrop now prints the filename of images where face detection failed
* Internal refactoring and more tests

## 0.3.1
### Changes
* Add `-r`, `--reject` flag to specify directory where the images that autocrop *couldn't* find a face in are directed to.
* Instead of having the target files copied then cropped, they are instead cropped and saved to their respective target folder.

## 0.3.0
### Changes
* Added support for padding (`padLeft`, etc.) in the CLI.

### Bugfix
* Fixed warp on crop for `-w` and `-h` values

## 0.2.0
### Changes
* Add `-o`, `--output` flag to specify directory where cropped images are to be dumped.
	- Error out if output folder set to current directory, i.e. `-o .`
	- If directory doesn't exist yet, create it.
	- If no face can be found in an image in batch, it is still copied over to `-o` folder.
	- If no output folder is added, ask for confirmation ([Y]/n), and destructively crop images in-place.
* Use `-i`, `--input` flags as synonyms for `-p` or `--path`: symmetrical in meaning to "output".
	- Is now standard nomenclature in documentation.
* `--input` or `--path` flag is now optional.
	- Standard behaviour without input folder is to non-recursively process all images in immediate folder, i.e. `-p .` as currently implemented.

### Breaks
* Removed all mentions of the hard-coded 'bkp' and 'crop' folders
* Calling autocrop without specifying an input path, i.e. `autocrop` does not look for the 'images' folder anymore.


tests/test_autocrop.py
# -*- coding: utf-8 -*-

"""Tests for autocrop"""

import io
import os
import shutil
import sys

import pytest
import cv2
import numpy as np
try:
    import mock
except ImportError:
    from unittest import mock

from autocrop.autocrop import gamma, crop, cli, main, size, confirmation

PY3 = sys.version_info[0] >= 3


@pytest.fixture()
def integration():
    # Setup
    path_i = "tests/test"
    path_o = "tests/crop"
    path_r = "tests/reject"
    shutil.copytree("tests/data", path_i)

    yield

    # Teardown
    shutil.rmtree(path_i)
    for path in [path_o, path_r]:
        try:
            shutil.rmtree(path)
        except OSError:
            pass


def test_gamma_brightens_image():
    """This function is so tightly coupled to cv2 it's probably useless.
    Still might flag cv2 or numpy boo-boos."""
    matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    expected = np.uint8([[15, 22, 27], [31, 35, 39], [42, 45, 47]])
    np.testing.assert_array_equal(gamma(img=matrix, correction=0.5), expected)


def test_crop_noise_returns_none():
    loc = "tests/data/noise.png"
    noise = cv2.imread(loc)
    assert crop(noise) is None


def test_obama_has_a_face():
    loc = "tests/data/obama.jpg"
    obama = cv2.imread(loc)
    assert len(crop(obama, 500, 500)) == 500


def test_size_140_is_valid():
    assert size(140) == 140


def test_size_0_not_valid():
    with pytest.raises(Exception) as e:
        size(0)
    assert "Invalid pixel" in str(e)


def test_size_million_not_valid():
    with pytest.raises(Exception) as e:
        size(1e6)
    assert "Invalid pixel" in str(e)


def test_size_asdf_gives_ValueError():
    with pytest.raises(Exception) as e:
        size("asdf")
    assert "ValueError" in str(e)


def test_size_minus_14_not_valid():
    with pytest.raises(Exception) as e:
        size(-14)
    assert "Invalid pixel" in str(e)


@mock.patch("autocrop.autocrop.input_path", lambda p: p)
@mock.patch("autocrop.autocrop.main")
def test_cli_no_args_means_cwd(mock_main):
    mock_main.return_value = None
    sys.argv = ["", "--no-confirm"]
    cli()
    args, _ = mock_main.call_args
    assert args == (".", None, None, 500, 500, 50, False, False, False, False)


@mock.patch("autocrop.autocrop.input_path", lambda p: p)
@mock.patch("autocrop.autocrop.main")
def test_cli_width_140_is_valid(mock_main):
    mock_main.return_value = None
    sys.argv = ["autocrop", "-w", "140", "--no-confirm"]
    assert mock_main.call_count == 0
    cli()
    assert mock_main.call_count == 1


def test_cli_invalid_input_path_errors_out():
    sys.argv = ["autocrop", "-i", "asdfasdf"]
    with pytest.raises(SystemExit) as e:
        cli()
    assert e.type == SystemExit
    assert "SystemExit" in str(e)


def test_cli_no_images_in_input_path():
    sys.argv = ["autocrop", "-i", "tests"]
    with pytest.raises(SystemExit) as e:
        cli()
    assert e.type == SystemExit
    assert "SystemExit" in str(e)


def test_cli_width_0_not_valid():
    sys.argv = ["autocrop", "-w", "0"]
    with pytest.raises(SystemExit) as e:
        cli()
    assert e.type == SystemExit
    assert "SystemExit" in str(e)


def test_cli_width_minus_14_not_valid():
    sys.argv = ["autocrop", "-w", "-14"]
    with pytest.raises(SystemExit) as e:
        cli()
    assert e.type == SystemExit
    assert "SystemExit" in str(e)


@pytest.mark.parametrize(
    "from_user, response, output",
    [
        (["x", "x", "No"], False, "Please respond with 'y' or 'n'\n" * 2),
        (["y"], True, ""),
        (["n"], False, ""),
        (["x", "y"], True, "Please respond with 'y' or 'n'\n"),
    ],
)
def test_confirmation_get_from_user(from_user, response, output):
    question = "Overwrite image files?"
    input_str = "autocrop.autocrop.compat_input"

    with mock.patch(input_str, lambda x: from_user.pop(0)):
        sio = io.StringIO if PY3 else io.BytesIO
        with mock.patch("sys.stdout", new_callable=sio):
            assert response == confirmation(question)
            assert output == sys.stdout.getvalue()


@mock.patch("autocrop.autocrop.main", lambda *args: None)
@mock.patch("autocrop.autocrop.confirmation")
def test_user_gets_prompted_if_no_output_is_given(mock_confirm):
    mock_confirm.return_value = False
    sys.argv = ["", "-i", "tests/data"]
    with pytest.raises(SystemExit) as e:
        assert mock_confirm.call_count == 0
        cli()
    assert mock_confirm.call_count == 1
    assert e.type == SystemExit


@mock.patch("autocrop.autocrop.main", lambda *args: None)
@mock.patch("autocrop.autocrop.confirmation")
def test_user_gets_prompted_if_output_same_as_input(mock_confirm):
    mock_confirm.return_value = False
    sys.argv = ["", "-i", "tests/data"]
    with pytest.raises(SystemExit) as e:
        assert mock_confirm.call_count == 0
        cli()
    assert mock_confirm.call_count == 1
    assert e.type == SystemExit


def test_main_overwrites_when_same_input_and_output(integration):
    sys.argv = ["", "--no-confirm", "-i", "tests/test", "-o", "tests/test"]
    cli()
    output_files = os.listdir(sys.argv[-1])
    assert len(output_files) == 11


@mock.patch("autocrop.autocrop.crop")
def test_main_overwrites_when_no_output(mock_crop, integration):
    mock_crop.return_value = None
    assert mock_crop.call_count == 0
    main("tests/test", None, None)
    assert mock_crop.call_count == 10


@mock.patch("autocrop.autocrop.main", lambda *args: None)
@mock.patch("autocrop.autocrop.output_path", lambda p: p)
@mock.patch("autocrop.autocrop.confirmation")
def test_user_does_not_get_prompted_if_output_d_is_given(mock_confirm):
    mock_confirm.return_value = False
    sys.argv = ["", "-i", "tests/data", "-o", "tests/crop"]
    assert mock_confirm.call_count == 0
    cli()
    assert mock_confirm.call_count == 0


@mock.patch("autocrop.autocrop.main", lambda *args: None)
@mock.patch("autocrop.autocrop.confirmation")
def test_user_does_not_get_prompted_if_no_confirm(mock_confirm):
    mock_confirm.return_value = False
    sys.argv = ["", "-i", "tests/data", "--no-confirm"]
    assert mock_confirm.call_count == 0
    cli()
    assert mock_confirm.call_count == 0


@mock.patch("autocrop.autocrop.crop", lambda *args: None)
def test_noface_files_copied_over_if_output_d_specified(integration):
    sys.argv = ["", "-i", "tests/test", "-o", "tests/crop"]
    cli()
    output_files = os.listdir(sys.argv[-1])
    assert len(output_files) == 10


def test_nofaces_copied_to_reject_d_if_both_reject_and_output_d(integration):
    sys.argv = [
        "",
        "-i",
        "tests/test",
        "-o",
        "tests/crop",
        "-r",
        "tests/reject",
    ]
    cli()
    output_files = os.listdir(sys.argv[-3])
    reject_files = os.listdir(sys.argv[-1])
    assert len(output_files) == 7
    assert len(reject_files) == 3


@mock.patch("autocrop.autocrop.confirmation", lambda *args: True)
def test_image_files_overwritten_if_no_output_dir(integration):
    sys.argv = ["", "-i", "tests/test"]
    cli()
    # We have the same number of files
    output_files = os.listdir(sys.argv[-1])
    assert len(output_files) == 11
    # Images with a face have been cropped
    shape = cv2.imread("tests/test/king.jpg").shape
    assert shape == (500, 500, 3)


autocrop/autocrop.py
# -*- coding: utf-8 -*-

from __future__ import print_function

import argparse
import cv2
import io
import numpy as np
import os
import shutil
import sys

from PIL import Image

from .__version__ import __version__
from .constants import (
    FIXEXP,
    MINFACE,
    GAMMA_THRES,
    GAMMA,
    QUESTION_OVERWRITE,
    CV2_FILETYPES,
    PILLOW_FILETYPES,
    CASCFILE,
)

COMBINED_FILETYPES = CV2_FILETYPES + PILLOW_FILETYPES
INPUT_FILETYPES = COMBINED_FILETYPES + [s.upper() for s in COMBINED_FILETYPES]

# Load XML Resource
d = os.path.dirname(sys.modules["autocrop"].__file__)
cascPath = os.path.join(d, CASCFILE)


# Load custom exception to catch a certain failure type
class ImageReadError(BaseException):
    pass


# Define simple gamma correction fn
def gamma(img, correction):
    img = cv2.pow(img / 255.0, correction)
    return np.uint8(img * 255)


def crop_positions(
    imgh,
    imgw,
    x,
    y,
    w,
    h,
    fheight,
    fwidth,
    facePercent,
    padUp,
    padDown,
    padLeft,
    padRight,
):
    # Check padding values
    padUp = 50 if (padUp is False or padUp < 0) else padUp
    padDown = 50 if (padDown is False or padDown < 0) else padDown
    padLeft = 50 if (padLeft is False or padLeft < 0) else padLeft
    padRight = 50 if (padRight is False or padRight < 0) else padRight

    # enfoce face percent
    facePercent = 100 if facePercent > 100 else facePercent
    facePercent = 50 if facePercent <= 0 else facePercent

    # Adjust output height based on Face percent
    height_crop = h * 100.0 / facePercent

    # Ensure height is within boundaries
    height_crop = imgh if height_crop > imgh else height_crop

    aspect_ratio = float(fwidth) / float(fheight)
    # Calculate width based on aspect ratio
    width_crop = aspect_ratio * float(height_crop)

    # Calculate padding by centering face
    xpad = (width_crop - w) / 2
    ypad = (height_crop - h) / 2

    # Calc. positions of crop
    h1 = float(x - (xpad * padLeft / (padLeft + padRight)))
    h2 = float(x + w + (xpad * padRight / (padLeft + padRight)))
    v1 = float(y - (ypad * padUp / (padUp + padDown)))
    v2 = float(y + h + (ypad * padDown / (padUp + padDown)))

    # Move crop inside photo boundaries
    while h1 < 0:
        h1 = h1 + 1
        h2 = h2 + 1
    while v1 < 0:
        v1 = v1 + 1
        v2 = v2 + 1
    while v2 > imgh:
        v2 = v2 - 1
        h2 = h2 - 1 * aspect_ratio
    while h2 > imgw:
        h2 = h2 - 1
        v2 = v2 - 1 / aspect_ratio

    return [int(v1), int(v2), int(h1), int(h2)]


def crop(
    image,
    fheight=500,
    fwidth=500,
    facePercent=50,
    padUp=False,
    padDown=False,
    padLeft=False,
    padRight=False,
):
    """Given a ndarray image with a face, returns cropped array.

    Arguments:
        - image, the numpy array of the image to be processed.
        - fwidth, the final width (px) of the cropped img. Default: 500
        - fheight, the final height (px) of the cropped img. Default: 500
        - facePercent, percentage of face height to image height. Default: 50
        - padUp, Padding from top
        - padDown, Padding to bottom
        - padLeft, Padding from left
        - padRight, Padding to right
    Returns:
        - image, a cropped numpy array

    ndarray, int, int -> ndarray
    """
    # Some grayscale color profiles can throw errors, catch them
    try:
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    except cv2.error:
        gray = image

    # Scale the image
    try:
        height, width = image.shape[:2]
    except AttributeError:
        raise ImageReadError
    minface = int(np.sqrt(height ** 2 + width ** 2) / MINFACE)

    # Create the haar cascade
    faceCascade = cv2.CascadeClassifier(cascPath)

    # ====== Detect faces in the image ======
    faces = faceCascade.detectMultiScale(
        gray,
        scaleFactor=1.1,
        minNeighbors=5,
        minSize=(minface, minface),
        flags=cv2.CASCADE_FIND_BIGGEST_OBJECT | cv2.CASCADE_DO_ROUGH_SEARCH,
    )

    # Handle no faces
    if len(faces) == 0:
        return None

    # Make padding from biggest face found
    x, y, w, h = faces[-1]

    pos = crop_positions(
        height,
        width,
        x,
        y,
        w,
        h,
        fheight,
        fwidth,
        facePercent,
        padUp,
        padDown,
        padLeft,
        padRight,
    )
    # Actual cropping
    image = image[pos[0]: pos[1], pos[2]: pos[3]]

    # Resize
    image = cv2.resize(image, (fwidth, fheight), interpolation=cv2.INTER_AREA)

    # ====== Dealing with underexposition ======
    if FIXEXP:
        # Check if under-exposed
        uexp = cv2.calcHist([gray], [0], None, [256], [0, 256])
        if sum(uexp[-26:]) < GAMMA_THRES * sum(uexp):
            image = gamma(image, GAMMA)
    return image


def open_file(input_filename):
    """Given a filename, returns a numpy array"""
    extension = os.path.splitext(input_filename)[1].lower()

    if extension in CV2_FILETYPES:
        # Try with cv2
        return cv2.imread(input_filename)
    if extension in PILLOW_FILETYPES:
        # Try with PIL
        with Image.open(input_filename) as img_orig:
            return np.asarray(img_orig)
    return None


def output(input_filename, output_filename, image):
    """Move the input file to the output location and write over it with the
    cropped image data."""
    if input_filename != output_filename:
        # Move the file to the output directory
        shutil.move(input_filename, output_filename)
    # Encode the image as an in-memory PNG
    img_png = cv2.imencode(".png", image)[1].tostring()
    # Read the PNG data
    img_new = Image.open(io.BytesIO(img_png))
    # Write the new image (converting the format to match the output
    # filename if necessary)
    img_new.save(output_filename)


def reject(input_filename, reject_filename):
    """Move the input file to the reject location."""
    if input_filename != reject_filename:
        # Move the file to the reject directory
        shutil.move(input_filename, reject_filename)


def main(
    input_d,
    output_d,
    reject_d,
    fheight=500,
    fwidth=500,
    facePercent=50,
    padUp=False,
    padDown=False,
    padLeft=False,
    padRight=False,
):
    """Crops folder of images to the desired height and width if a face is found

    If input_d == output_d or output_d is None, overwrites all files
    where the biggest face was found.

    Args:
        input_d (str): Directory to crop images from.
        output_d (str): Directory where cropped images are placed.
        reject_d (str): Directory where images that cannot be cropped are
                        placed.
        fheight (int): Height (px) to which to crop the image.
                       Default: 500px
        fwidth (int): Width (px) to which to crop the image.
                       Default: 500px
        facePercent (int) : Percentage of face from height,
                       Default: 50
    Side Effects:
        Creates image files in output directory.

    str, str, (int), (int) -> None
    """
    reject_count = 0
    output_count = 0
    input_files = [
        os.path.join(input_d, f)
        for f in os.listdir(input_d)
        if any(f.endswith(t) for t in INPUT_FILETYPES)
    ]
    if output_d is None:
        output_d = input_d
    if reject_d is None and output_d is None:
        reject_d = input_d
    if reject_d is None:
        reject_d = output_d

    # Guard against calling the function directly
    input_count = len(input_files)
    assert input_count > 0

    # Main loop
    for input_filename in input_files:
        basename = os.path.basename(input_filename)
        output_filename = os.path.join(output_d, basename)
        reject_filename = os.path.join(reject_d, basename)

        input_img = open_file(input_filename)
        image = None

        # Attempt the crop
        try:
            image = crop(
                input_img,
                fheight,
                fwidth,
                facePercent,
                padUp,
                padDown,
                padLeft,
                padRight,
            )
        except ImageReadError:
            print("Read error:       {}".format(input_filename))
            continue

        # Did the crop produce an invalid image?
        if isinstance(image, type(None)):
            reject(input_filename, reject_filename)
            print("No face detected: {}".format(reject_filename))
            reject_count += 1
        else:
            output(input_filename, output_filename, image)
            print("Face detected:    {}".format(output_filename))
            output_count += 1

    # Stop and print status
    print(
        "{} input files, {} faces cropped, {} rejected".format(
            input_count, output_count, reject_count
        )
    )


def input_path(p):
    """Returns path, only if input is a valid directory"""
    no_folder = "Input folder does not exist"
    no_images = "Input folder does not contain any image files"
    p = os.path.abspath(p)
    if not os.path.isdir(p):
        raise argparse.ArgumentTypeError(no_folder)
    filetypes = set(os.path.splitext(f)[-1] for f in os.listdir(p))
    if not any(t in INPUT_FILETYPES for t in filetypes):
        raise argparse.ArgumentTypeError(no_images)
    else:
        return p


def output_path(p):
    """Returns path, if input is a valid directory name.
    If directory doesn't exist, creates it."""
    p = os.path.abspath(p)
    if not os.path.isdir(p):
        os.makedirs(p)
    return p


def size(i):
    """Returns valid only if input is a positive integer under 1e5"""
    error = "Invalid pixel size"
    try:
        i = int(i)
    except TypeError:
        raise argparse.ArgumentTypeError(error)
    if i > 0 and i < 1e5:
        return i
    else:
        raise argparse.ArgumentTypeError(error)


def compat_input(s=""):
    """Compatibility function to permit testing for Python 2 and 3"""
    try:
        return raw_input(s)
    except NameError:
        # Py2 raw_input() renamed to input() in Py3
        return input(s)  # lgtm[py/use-of-input]


def confirmation(question, default=True):
    """Ask a yes/no question via standard input and return the answer.

    If invalid input is given, the user will be asked until
    they acutally give valid input.

    Args:
        question(str):
            A question that is presented to the user.
        default(bool|None):
            The default value when enter is pressed with no value.
            When None, there is no default value and the query
            will loop.
    Returns:
        A bool indicating whether user has entered yes or no.

    Side Effects:
        Blocks program execution until valid input(y/n) is given.
    """
    yes_list = ["yes", "y"]
    no_list = ["no", "n"]

    default_dict = {  # default => prompt default string
        None: "[y/n]",
        True: "[Y]/n",
        False: "y/[N]",
    }

    default_str = default_dict[default]
    prompt_str = "%s %s " % (question, default_str)

    while True:
        choice = compat_input(prompt_str).lower()

        if not choice and default is not None:
            return default
        if choice in yes_list:
            return True
        if choice in no_list:
            return False

        notification_str = "Please respond with 'y' or 'n'"
        print(notification_str)


def parse_args(args):
    help_d = {
        "desc": "Automatically crops faces from batches of pictures",
        "input": """Folder where images to crop are located. Default:
                     current working directory""",
        "output": """Folder where cropped images will be moved to.

                      Default: current working directory, meaning images are
                      cropped in place.""",
        "reject": """Folder where images that could not be cropped will be
                       moved to.

                      Default: current working directory, meaning images that
                      are not cropped will be left in place.""",
        "width": "Width of cropped files in px. Default=500",
        "height": "Height of cropped files in px. Default=500",
        "y": "Bypass any confirmation prompts",
        "facePercent": "Percentage of face to image height",
        "padUp": "Add padding up to face cropped",
        "padDown": "Add padding down to face cropped",
        "padLeft": "Add padding left to face cropped",
        "padRight": "Add padding right to face cropped",
    }

    parser = argparse.ArgumentParser(description=help_d["desc"])
    parser.add_argument(
        "-i", "--input", default=".", type=input_path, help=help_d["input"]
    )
    parser.add_argument(
        "-o",
        "--output",
        "-p",
        "--path",
        type=output_path,
        default=None,
        help=help_d["output"],
    )
    parser.add_argument(
        "-r", "--reject", type=output_path, default=None, help=help_d["reject"]
    )
    parser.add_argument(
        "-w", "--width", type=size, default=500, help=help_d["width"]
    )
    parser.add_argument(
        "-H", "--height", type=size, default=500, help=help_d["height"]
    )
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version="%(prog)s version {}".format(__version__),
    )
    parser.add_argument("--no-confirm", action="store_true", help=help_d["y"])
    parser.add_argument(
        "--padUp", type=size, default=False, help=help_d["padUp"]
    )
    parser.add_argument(
        "--padDown", type=size, default=False, help=help_d["padDown"]
    )
    parser.add_argument(
        "--padLeft", type=size, default=False, help=help_d["padLeft"]
    )
    parser.add_argument(
        "--padRight", type=size, default=False, help=help_d["padRight"]
    )
    parser.add_argument(
        "--facePercent", type=size, default=50, help=help_d["facePercent"]
    )

    return parser.parse_args()


def cli():
    args = parse_args(sys.argv[1:])
    if not args.no_confirm:
        if args.output is None or args.input == args.output:
            if not confirmation(QUESTION_OVERWRITE):
                sys.exit()
    if args.input == args.output:
        args.output = None
    print("Processing images in folder:", args.input)
    main(
        args.input,
        args.output,
        args.reject,
        args.height,
        args.width,
        args.facePercent,
        args.padUp,
        args.padDown,
        args.padLeft,
        args.padRight,
    )
